src/wui/constructionsitewindow.cc

```diff
@@ -127,97 +127,87 @@ void ConstructionSiteWindow::init(bool avoid_fastclick, bool workarea_preview_wa
-			// We use the ProductionSiteDescr to get the correct order for the input queues
-			upcast(const Widelands::ProductionSiteDescr, prodsite, &construction_site->building());
-			assert(prodsite != nullptr);
-			assert(ps->ware_queues.size() == prodsite->input_wares().size());
-			ensure_box_can_hold_input_queues(settings_box);
-			for (const auto& pair : prodsite->input_wares()) {
-				InputQueueDisplay* queue = new InputQueueDisplay(
-				   &settings_box, *ibase(), *construction_site, Widelands::wwWARE, pair.first);
-				settings_box.add(queue, UI::Box::Resizing::kFullSize);
-				cs_ware_queues_.push_back(queue);
-			}
-			assert(ps->worker_queues.size() == prodsite->input_workers().size());
-			for (const auto& pair : prodsite->input_workers()) {
-				InputQueueDisplay* queue = new InputQueueDisplay(
-				   &settings_box, *ibase(), *construction_site, Widelands::wwWORKER, pair.first);
-				settings_box.add(queue, UI::Box::Resizing::kFullSize);
-				cs_ware_queues_.push_back(queue);
-			}
-			if (upcast(Widelands::TrainingsiteSettings, ts, ps)) {
-				cs_soldier_capacity_ = new ConstructionSoldierCapacityBox(
-				   &settings_box, ts->desired_capacity, 0, ts->max_capacity, can_act);
-				settings_box.add(cs_soldier_capacity_, UI::Box::Resizing::kAlign, UI::Align::kCenter);
-				settings_box.add_space(8);
-				cs_soldier_capacity_->changed.connect([this]() {
-					if (game_) {
-						game_->send_player_change_soldier_capacity(
-						   *construction_site_.get(ibase()->egbase()),
-						   cs_soldier_capacity_->get_current());
-					} else {
-						NEVER_HERE();  // TODO(Nordfriese / Scenario Editor): implement
-					}
-				});
-			}
-			cs_stopped_ =
-			   new UI::Checkbox(&settings_box, UI::PanelStyle::kWui, Vector2i::zero(), _("Stopped"),
-			                    _("Stop this building’s work after completion"));
-			cs_stopped_->clickedto.connect([this, ps](bool stop) {
-				if (stop != ps->stopped) {
-					if (game_) {
-						game_->send_player_start_stop_building(
-						   *construction_site_.get(ibase()->egbase()));
-					} else {
-						NEVER_HERE();  // TODO(Nordfriese / Scenario Editor): implement
-					}
-				}
-			});
-			settings_box.add(cs_stopped_, UI::Box::Resizing::kAlign, UI::Align::kCenter);
-			settings_box.add_space(6);
-			cs_stopped_->set_enabled(can_act);
-		} break;
-		case Widelands::MapObjectType::MILITARYSITE: {
-			upcast(Widelands::MilitarysiteSettings, ms, construction_site->get_settings());
+}
+void ConstructionSiteWindow::build_settings_tab(Widelands::ConstructionSite* construction_site) {
+	const bool can_act = ibase()->can_act(construction_site->owner().player_number());
+	// Create the settings. Since we don't access an actual building, we create
+	// a simplified faksimile of the later building window that contains only
+	// the relevant options.
+	bool nothing_added = false;
+	UI::Box& settings_box = *new UI::Box(get_tabs(), UI::PanelStyle::kWui, 0, 0, UI::Box::Vertical);
```

Comment:
This should better be a `std::unique_ptr<UI::Box>` which is `release()`d in the call to `get_tabs()->add()`. Same for the boxes in the other functions, to avoid memory leaks if an exception happens or in the case that `nothing_added` is `true`.

Reply:
let's do this another time :)